### PR TITLE
Implementation of TDD and Validator

### LINKED ISSUE #1: Adopt TDD approach: Write tests before implementation
Since we have a clear idea of the implementation, we should start by writing tests to follow a Test-Driven Development (TDD) approach.

**Action items:**
- Initiate the development cycle by writing tests for each new feature or improvement before the actual code implementation. 
- Create automated tests that use curl commands to call endpoints and validate responses against expected outputs.
- Integrate these automated checks into the CI/CD pipeline if possible.

This will ensure code quality and correctness from the start.

### PR DESCRIPTION
**Pull request by JorgeSaicoski**
This pull request introduces a comprehensive foundation for implementing the Transaction Ledger Service using a Test-Driven Development (TDD) approach. It adds a complete TDD test suite, supporting documentation, CI/CD integration, and initial project scaffolding. The changes ensure that all requirements and edge cases are covered by tests before any service logic is implemented, setting up a robust workflow for future development.

**Key changes include:**

---

**TDD Test Suite and Documentation:**

* Added a full suite of bash-based API tests covering all endpoints and edge cases, including scripts for creating, retrieving, listing transactions, and checking balances, as well as shared configuration and examples. Supporting documentation details test cases and usage. (`tests/`, `tests/README.md`, `tests/TEST_SPECIFICATION.md`, `TDD_SUMMARY.md`, `GETTING_STARTED.md`, [[1]](diffhunk://#diff-64f5e836e5ad8ed904b012122f281374e76d5d2993795c6b40aac565fab96d0eR1-R136) [[2]](diffhunk://#diff-d0243df16bd9ebcc544cb26394cf6ec5cc2be67bdca0a44c51457c0baf94b1c8R1-R324)
* Updated the main `README.md` with a new Testing section outlining TDD principles, test coverage, and CI/CD integration. (`README.md`, [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R164-R195) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R204)

---

**Project Structure and Architecture:**

* Added `PROJECT_STRUCTURE.md` describing the service's architecture, design principles, data flow, API endpoints, database schema, error handling, and scalability considerations. (`PROJECT_STRUCTURE.md`, [PROJECT_STRUCTURE.mdR1-R238](diffhunk://#diff-ca38625d9fdefc0c6ed46e68c5e068ff00aa09314ce7aafb1039b7722a48c00bR1-R238))
* Created an example `.env.example` file with environment variables for database connection, server port, and allowed origin. (`.env.example`, [.env.exampleR1-R8](diffhunk://#diff-a3046da0d15a27e89f2afe639b25748a7ad4d9290af3e7b1b6c1a5533c8f0a8cR1-R8))
* Added an empty `Makefile` as a placeholder for future build commands. (`Makefile`, [MakefileR1](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R1))

---

**CI/CD and Implementation Scaffolding:**

* Introduced a GitHub Actions workflow (`.github/workflows/tests.yml`) to automate running the test suite on push and pull requests, including service container setup and test execution steps (currently commented until implementation is ready). (`.github/workflows/tests.yml`, [.github/workflows/tests.ymlR1-R93](diffhunk://#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957fR1-R93))
* Added a stub `main.go` as the entry point for the future Go implementation, with comments outlining the steps to build the service. (`cmd/server/main.go`, [cmd/server/main.goR1-R28](diffhunk://#diff-53cd25fc908d5e63cc8675afbd4ff0910572a985a7cfec27f4460eb674f3b98eR1-R28))

---

These changes collectively establish a strong foundation for building the ledger service with confidence, ensuring that implementation will be guided and validated by comprehensive, automated tests from the outset.
This pull request lays the foundation for the Transaction Ledger Service project by establishing its architecture, testing strategy, and CI/CD integration. The most significant changes are the addition of comprehensive documentation detailing the system design and TDD approach, the creation of a full test suite before implementation, and the setup of environment and workflow files for future development and automated testing.

**Documentation and Architecture**

* Added `PROJECT_STRUCTURE.md` with a thorough description of the service's architecture, design principles, security model, API endpoints, database schema, error handling, scalability plans, and folder structure.
* Updated `README.md` to include a new testing section outlining TDD practices, test coverage, and CI/CD integration. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R164-R195) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R204)

**Test-Driven Development (TDD) and Test Suite**

* Added `TDD_SUMMARY.md` summarizing the TDD approach, listing all test scripts and documentation files, and describing comprehensive test coverage for all API endpoints and edge cases.

**CI/CD and Environment Setup**

* Added `.github/workflows/tests.yml` to configure GitHub Actions for automated API integration tests, including service setup and placeholders for future implementation steps.
* Added `.env.example` with example environment variables for database connection, server port, and allowed origin to standardize development and testing environments.

**Project Bootstrap**

* Added a placeholder `cmd/server/main.go` file with a skeleton main function and TODOs for implementing the service entry point and HTTP routes.
* Created an empty `Makefile` for future build and development automation.

### CODE REVIEW

File: PROJECT_STRUCTURE.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    user_id VARCHAR(100) NOT NULL,
+    amount DECIMAL(19,4) NOT NULL,
+    currency VARCHAR(20) NOT NULL,
+    created_at TIMESTAMP NOT NULL DEFAULT NOW()
+);
+
+CREATE INDEX idx_user_currency ON transactions(user_id, currency);
+CREATE INDEX idx_created_at ON transactions(created_at);
+```
+
```
[the remark this answers is not in the document]
Comment (JorgeSaicoski) [reply PROJECT_STRUCTURE.md:125]: Actually we will use the minimal division of the coin. For example, for BRL we will store the "centavos"
For USD we will use the "cents".
UYU we will use "peso"

File: internal/validator/transaction_validator.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+type TransactionValidator struct {
+	currencyRegex *regexp.Regexp
+	uuidRegex     *regexp.Regexp
+}
+
+// NewTransactionValidator creates a new validator instance
+func NewTransactionValidator() *TransactionValidator {
+	return &TransactionValidator{
+		currencyRegex: regexp.MustCompile(`^[a-z0-9_]+$`),
+		uuidRegex:     regexp.MustCompile(`^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$`)}
```
[the remark this answers is not in the document]
Comment (JorgeSaicoski) [reply internal/validator/transaction_validator.go:33]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/JorgeSaicoski/ledger-service/pull/4#discussion_r2637448043)

We must update the documentation to not accept upper case. And be sure that the tests are testing upper case as not accept.

The client must send it as low case.

File: tests/README.md [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+# Test only transaction listing
+./tests/test_list_transactions.sh
+```
+
+## Test Coverage
+
+### POST /transactions (10 tests)
+- ✓ Create transaction with valid data and allowed origin (201)
+- ✓ Create transaction with disallowed origin returns 403
+- ✓ Create transaction without Origin header returns 403
```
[the remark this answers is not in the document]
Comment (JorgeSaicoski) [reply tests/README.md]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/JorgeSaicoski/ledger-service/pull/4#discussion_r2637448046)

We must update the documentation. We will not have middleware here, we will trust in the gateway to receive only allowed request.

File: internal/models/transaction.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package models
+
+import "time"
+
+// Transaction represents a ledger transaction
+type Transaction struct {
+	ID        string    `json:"id"`
+	UserID    string    `json:"user_id"`
+	Amount    int       `json:"amount"`
```
[the remark this answers is not in the document]
Comment (JorgeSaicoski) [reply internal/models/transaction.go:9]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/JorgeSaicoski/ledger-service/pull/4#discussion_r2637448050)

Actually we will use the minimal division of the coin. For example, for BRL we will store the "centavos"
For USD we will use the "cents".
UYU we will use "peso"

So we need to update the documentation, test and whatever more that the code is expecting float.
We can't store or calculate float number, we will use int.
